keps/draft-20180426-namespace-template.md

```diff
@@ -0,0 +1,513 @@
+
+When multiple templates matches the same namespace, 
+the templates are concatenated in the alphabetic order of the name of `NamespaceTemplate`.
+
+A `NamespaceTemplate` can be disabled by an explicit annotation:
+
+```
+policy/namespace-template-apply=disable
+```
+
+#### Schema validation
+
+As `NamespaceTemplate` is defined as CRD (Custom Resource Definition),
+currently there's no effective way to validate the schema of CR (Custom Resource)
+if it contains complex types, or encapsulating existing object types.
+This problem is out-of-scope here.
+As a result, the `templates` are not validated during the creation of `NamespaceTemplate`,
+and it will fail at population time (via `kubectl apply`) if anything is written incorrectly.
+The failure will be reported as Kubernetes Events in the target namespace.
+
+An alternative is to create a template namespace,
+and put all these objects into the namespace as real resources.
+When a new namespace is created, the objects in the template namespace are copied over.
+As real objects are created, they are effective, and may cause side effects,
+though within known object types, there's no impact of the effectiveness of these objects.
+
+There are other alternatives that the objects can be defined somewhere else from `NamespaceTemplate`.
+The implementation can be pluggable (e.g. adding a `source` property) to support different sources of templates.
+While the initial proposal will focus on inlined objects.
+
+The MVP scope covers the inlining templates only.
+
+#### Opt-out
+
+Cluster administrator is able to opt-out some namespaces from being populated by the controller.
+`kube-system` is always opt-out.
+For other namespaces, there are two options to opt-out from automated population:
+
+1. Put a label that is excluded in the `labelSelector` of the `NamespaceTemplate`
+2. Annotate the namespace with `policy/namespace-template-opt-out=true`
+
+Note: the objects previously populated by the controller will be left as-is in the namespace when it's opted out.
+
+Normally, there's no _exclusive_ condition in `labelSelector` of the `NamespaceTemplate`, so option 2 is recommended.
+
+Some ephemeral namespaces created by privileged controllers may need to be opt-out
+as they are not accessibly by users and will be managed completely by the controllers.
+As these controllers don't have knowledge about `NamespaceTemplate`, 
+or it's almost impossible to customize the way how they create namespaces,
+an automated opt-out mechanism is required.
+The mechanism is based on the service account running the controllers.
+With a `MutatingAdmissionWebhook`, 
+a namespace created by these controllers will be automatically annotated for opt-out.
+This mechanism is out of MVP scope.
+
+#### Self-serviced namespace creation
+
+Self-serviced namespace creation allows certain users/groups to create namespaces
```

Comment:
I don't think this proposal as written can correctly protect against this, so it either needs to be dropped, split, or a lot more work put into it.

Reply:
@smarterclayton What is the "this" you're referring to where you say "protect against this"? I think you are referring to something that is not in this sentence, but I wasn't sure exactly what.

Certainly people are doing self-service namespace creation today on stock upstream Kubernetes clusters, so I assume you are referring to doing that with some additional guarantees. Can you specify which guarantees you think this proposal does not cover? The one you have mentioned elsewhere is quota on namespace creation. Are there others?